internal/factsengine/gatherers/systemd.go

```diff
@@ -63,43 +63,58 @@ func NewSystemDGatherer(conn DbusConnector, initialized bool) *SystemDGatherer {
 	}
 }
 
-func (g *SystemDGatherer) Gather(_ context.Context, factsRequests []entities.FactRequest) ([]entities.Fact, error) {
-	facts := []entities.Fact{}
+func (g *SystemDGatherer) Gather(ctx context.Context, factsRequests []entities.FactRequest) ([]entities.Fact, error) {
 	log.Infof("Starting %s facts gathering process", SystemDGathererName)
 
 	if !g.initialized {
-		return facts, &SystemDNotInitializedError
+		return nil, &SystemDNotInitializedError
 	}
 
-	services := []string{}
-	for _, factReq := range factsRequests {
-		if len(factReq.Argument) == 0 {
-			continue
-		}
-		services = append(services, completeServiceName(factReq.Argument))
-	}
+	results := make(chan []entities.Fact)
+	errors := make(chan error)
 
-	ctx := context.Background()
+	go func() {
```

Comment:
question: why do we need a thread?
I would suppose that the `ListUnitsByNamesContext` function returns immediately when the context is cancelled. If that's the case we can simply return the error it returns.
We basically want to return when the context is cancelled, we don't really care even which error we return, we are simply cancelling the whole process.

At omst you can check if the context is done after running the `ListUnitsByNamesContext` if you want to have a different error

Reply:
It's a bit overkill, indeed. The refactoring _per se_ has two benefits:
1. don't wait for `ListUnitsByNamesContext` to finish before closing the gathering process
2. have an explicit function response

If you think both are of not much value, we can rollback

Reply:
> don't wait for ListUnitsByNamesContext to finish before closing the gathering process

I think we would wait the same either way. I expect `ListUnitsByNamesContext` returning right away when the context we pass is cancelled, and we return from that function if this happens. If `ListUnitsByNamesContext` doesn't work that way it is a bit of a bummer, as we don't even need to pass the context to it (it is not doing anything).
In fact, we can enter in a race condition and have a different result depending on the `ctx.Done` of the `ListUnitsByNamesContext` or the one you are using in the `select` reacts first.

Reply:
Makes sense